### PR TITLE
Module 4

### PR DESCRIPTION
**Pull request by cpujades**
@gsantam @guillerbf 

The notebook is where most of the testing has been done (Random Forest, GradientBoosting...) using ROC and PR curves.

The train.py file includes the necessary functions to train the model, as well as the function asked in the TDD, `handler_fit`.

The utils.py includes helper functions.

The inference.py includes the function to make predictions over new data, `handler_predict`.

### CODE REVIEW

File: src/module_4/notebook.md [hunk 1]
```diff
@@ -0,0 +1,2102 @@
...
+categorical_cols = ["product_type", "vendor"]
+binary_cols = ["ordered_before", "abandoned_before", "active_snoozed", "set_as_regular"]
+numerical_cols = [col for col in features_cols if col not in categorical_cols + binary_cols]
+```
+
+
+```python
+order_size = df.groupby("order_id").outcome.sum().reset_index()
+big_orders = order_size.loc[order_size["outcome"] >= 5, "order_id"]
+df_selected = df.loc[lambda x: x.order_id.isin(big_orders)]
```
Comment (gsantam) [inline src/module_4/notebook.md:256]: you don't need lambda (and probably is gonna be slower), instead `df_selected = df[df["order_id"].isin(big_orders)]`

Comment (gsantam) [reply src/module_4/notebook.md:256]: also put a `.copy()`at the end of the df (`df_selected = df[df["order_id"].isin(big_orders)].copy()`) to avoid future warnings

File: src/module_4/notebook.md [hunk 2]
```diff
@@ -0,0 +1,2102 @@
...
+plt.title("Daily Orders");
+```
+
+
+    
+![png](notebook_files/notebook_9_0.png)
+    
+
+
+To avoid information leakage, like having the same orders in train/test, we make a temporal split. Also, due to the strong temporal component, it makes sense to split data in a chronological way and not randomly.
```
Comment (gsantam) [inline src/module_4/notebook.md:676]: This evaluation schema is not totally incorrect but I would consider: 
* make the 70% 20% 10% split over days instead of over orders (bear in mind that in the train period you are NOT supposed to know how many orders you are gonna have in the future, while with the schema that you are proposing you are implicitly assuming this number, this is not true if you just consider number of days)
* Have a rolling window where you keep moving the train / test to the right and you train and evaluate in different months to see how the model behaves in different periods

File: src/module_4/notebook.md [hunk 3]
```diff
@@ -0,0 +1,2102 @@
...
+
+    Train since:  2020-10-05 00:00:00
+    Train until:  2021-02-04 00:00:00
+    Val until:  2021-02-22 00:00:00
+    Test until:  2021-03-03 00:00:00
+
+
+
+```python
+train_df = df_selected[df_selected.order_date <= train_val_cutoff]
```
Comment (gsantam) [inline src/module_4/notebook.md:700]: as a health check print the min and max of date for each data frame

File: src/module_4/notebook.md [hunk 4]
```diff
@@ -0,0 +1,2102 @@
...
+    Returns the ROC and PR curves AUC score for a given set of true labels and predicted probabilities.
+
+    Args:
+        y_test (array-like): True labels.
+        y_preds (array-like): Predicted probabilities or scores.
+
+    Returns:
+        Tuple (float, float): ROC AUC score, PR curve AUC score.
+    """
+    fpr, tpr, _ = roc_curve(y_test, y_pred)
```
Comment (gsantam) [inline src/module_4/notebook.md:766]: this code is duplicated with some that you have in `plot_roc_and_pr_curves`. Keep it in just one place

File: src/module_4/notebook.md [hunk 5]
```diff
@@ -0,0 +1,2102 @@
...
+    roc_auc = auc(fpr, tpr)
+    precision_, recall_, _ = precision_recall_curve(y_test, y_pred)
+    pr_auc = auc(recall_, precision_)
+    return roc_auc, pr_auc
+```
+
+### Baseline
+
+We first try our baseline without any ML involved, just using a single column, `global_popularity` to predict the outcome.
```
Comment (gsantam) [inline src/module_4/notebook.md:775]: good idea to always use a baseline!

Comment (gsantam) [reply src/module_4/notebook.md:775]: however... is "total popularity" calculated over the WHOLE data set? does this mean that you are leaking data from the future in the past? I would double check that because otherwise this baseline is not totally correct

File: src/module_4/notebook.md [hunk 6]
```diff
@@ -0,0 +1,2102 @@
...
+roc_auc_scores_val = []
+
+fig1, ax1 = plt.subplots(1, 2, figsize=(12, 5))
+fig1.suptitle("Train metrics")
+
+fig2, ax2 = plt.subplots(1, 2, figsize=(12, 5))
+fig2.suptitle("Validation metrics")
+
+lrs = [
+    make_pipeline(StandardScaler(), LogisticRegression(penalty='l2', C=1e-6)),
```
Comment (gsantam) [inline src/module_4/notebook.md:824]: how have you chosen these params? from the prev notbeooks? they seem extremely low

File: src/module_4/notebook.md [hunk 7]
```diff
@@ -0,0 +1,2102 @@
...
+    make_pipeline(StandardScaler(), LogisticRegression(penalty='l1', C=1e-4, solver='saga')),
+]
+
+names = ["Ridge 1e-6", "Lasso 1e-4"]
```
Comment (gsantam) [inline src/module_4/notebook.md:828]: nit, it's better to make a dictionary: ```lrs = {
    "Ridge 1e-6":make_pipeline(StandardScaler(), LogisticRegression(penalty='l2', C=1e-6)),
    "Lasso 1e-4":make_pipeline(StandardScaler(), LogisticRegression(penalty='l1', C=1e-4, solver='saga')),
}```

Comment (gsantam) [reply src/module_4/notebook.md:828]: good that you put the standardscaler!

File: src/module_4/notebook.md [hunk 8]
```diff
@@ -0,0 +1,2102 @@
...
+    train_proba = lr.predict_proba(X_train[train_cols])[:, 1]
+    plot_roc_and_pr_curves(name, y_pred=train_proba, y_test=y_train, figure=(fig1, ax1))
+    
+    val_proba = lr.predict_proba(X_val[train_cols])[:, 1]
+    plot_roc_and_pr_curves(name, y_pred=val_proba, y_test=y_val, figure=(fig2, ax2))
+    auc_scores_val = auc_score(y_val, val_proba)
+    roc_auc_scores_val.append(auc_scores_val[0])
+    pr_auc_score_val.append(auc_scores_val[1])
+    
+    coeffs = lr.named_steps["logisticregression"].coef_[0]
```
Comment (gsantam) [inline src/module_4/notebook.md:840]: maybe you can put this into a function?

File: src/module_4/notebook.md [hunk 9]
```diff
@@ -0,0 +1,2102 @@
...
+
+
+    
+![png](notebook_files/notebook_24_1.png)
+    
+
+
+
+    
+![png](notebook_files/notebook_24_2.png)
```
Comment (gsantam) [inline src/module_4/notebook.md:863]: Again, please let's be quite skeptical about this global_popularity feature. If it was calculated over the whole dataset it's clearly leaking

File: src/module_4/notebook.md [hunk 10]
```diff
@@ -0,0 +1,2102 @@
...
+    plot_roc_and_pr_curves(name, y_pred=train_proba, y_test=y_train, figure=(fig1, ax1))
+    
+    val_proba = lr.predict_proba(X_val[train_cols])[:, 1]
+    plot_roc_and_pr_curves(name, y_pred=val_proba, y_test=y_val, figure=(fig2, ax2))
+    auc_scores_val = auc_score(y_val, val_proba)
+    roc_auc_scores_val.append(auc_scores_val[0])
+    pr_auc_score_val.append(auc_scores_val[1])
+    
+    coeffs = lr.named_steps["logisticregression"].coef_[0]
+    feat_importance = pd.DataFrame({"feature": train_cols, "importance": np.abs(coeffs)})
```
Comment (gsantam) [inline src/module_4/notebook.md:841]: why abs? It is interesting to know the direction of the importance and you are always gonna know the absolute value of the importance based on the length of the bar

File: src/module_4/notebook.md [hunk 11]
```diff
@@ -0,0 +1,2102 @@
...
+n_estimtors = [1, 10, 50, 100]
+max_depths = [2, 5, 7, 9]
+
+hyperparameters = []
+
+for d in max_depths:
+    for n in n_estimtors:
+        rf = RandomForestClassifier(n_estimators=n, max_depth=d, random_state=0)
+        rf.fit(X_train[train_cols], y_train)
+        train_proba = rf.predict_proba(X_train[train_cols])[:, 1]
```
Comment (gsantam) [inline src/module_4/notebook.md:953]: it seems that you can put this into a function to not repeat each time

File: src/module_4/notebook.md [hunk 12]
```diff
@@ -0,0 +1,2102 @@
...
+
+n_estimtors = [1, 10, 50, 100]
+max_depths = [2, 5, 7, 9]
+learnning_rate = [0.001, 0.01, 0.1]
+
+hyper_estimators = []
+hyper_depths = []
+hyper_lr = []
+
+for n in n_estimtors:
```
Comment (gsantam) [inline src/module_4/notebook.md:1299]: You can use: https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.ParameterGrid.html to make these loops more elegant

File: src/module_4/notebook.md [hunk 13]
```diff
@@ -0,0 +1,2102 @@
...
+      <td>0.844664</td>
+      <td>0.184307</td>
+    </tr>
+  </tbody>
+</table>
+</div>
+
+
+
+ Focusing on the AUC scores, we can see that RF is ahead of GBT, although not by much. If we check in the chart above, we can see that both ROC Curves are very similar. However, the thing changes when we focus on the PR Curve. With the RF ahead until a recall of ~0.4, from there, the two lines are similar.
```
Comment (gsantam) [inline src/module_4/notebook.md:2102]: The idea would be know to evaluate the best model on test to see the non overfitted definitive metric

**Review by gsantam — APPROVED** [review]
Good work, well structured, clean code and nice conclusions. Drop some comments about validation and data leakage which may be relevant
